### PR TITLE
PLUGIN Markdown

### PR DESCRIPTION
**Pull request by ArthurHoaro**
Parse link description in Markdown (HTML) before rendering.

  * hard remove of Shaarli's HTML before parsing.
  * Using Parsedown <https://github.com/erusev/parsedown> PHP lib.
  * fixes a bug where description links with a redirector aren't using urlencode() + coding style.

### CODE REVIEW

**Comment by ArthurHoaro** [comment]
Updated.

TODO:
  * [ ] add unit tests.
  * [ ] move url redirector fix to #355 

HTML cleanup isn't really pretty, I believe a plugin shouldn't alter core code to work properly.

Otherwise feel free to review and test it.

Side question, should we allow image tags in shaare descriptions? I'm not sure it's really relevant.

File: application/Utils.php [hunk 1]
```diff
@@ -168,3 +168,35 @@ function is_session_id_valid($sessionId)
...
+    // Redirector is , urlencode the final URL.
+    return preg_replace(
+        $regex . 'e',
+        "'<a href=\"". $redirector ."'. urlencode('$1') .'\" rel=\"nofollow\">$1</a>'",
+        $text
+    );
+}
+
+// This function inserts &nbsp; where relevant so that multiple spaces are properly displayed in HTML
+// even in the absence of <pre>  (This is used in description to keep text formatting)
```
Comment (ArthurHoaro) [inline application/Utils.php]: Comments.

**Comment by ArthurHoaro** [comment]
Blocked by #355.

**Comment by ArthurHoaro** [comment]
Updated. Almost good. 

Need:

  - [ ] Commits organization.
  - [ ] A bit of CSS improvement.

**Comment by alexisju** [comment]
CSS : 
- Je ne suis pas sur qu'il faille spécifier une famille de police (en tout cas, par défaut ou pour les p), laissons cela aux templates. 
 - Idem pour les couleurs de liens.
- idem pour le max-width est-il nécessaire ? (ou alors indiquer une largeur relative)

Sinon, le plugin semble très fonctionnel :-)

**Comment by nicolasdanelon** [comment]
@alexisju thanks for comments in french! Are very useful. Shaarli markdown plugin need a default font. That font can be easy override in a custom. Just need to call the new css later or use !important. 
I'm ok with margins, max-width need to be removed but colors.. same that font family. Can be override.

**Comment by alexisju** [comment]
I'm not sure that a plugin needs to change general layout of Shaarli. I think it should 'only' enhance capabilities but not to be in conflict with shaarli default template or with other templates. Otherwise, for my point of view, it is *breaking* things. 

Why not using *inherit* value for general styling (fonts, colors,...) ?

Honestly, it will be a lot easier to do that than put dozen of *!important* into shaarli.css or other css.

**Comment by ArthurHoaro** [comment]
Thanks for your feedback guys. 

You're right, I'll cleanup this file to comply to Shaarli's theme as much as possible.

**Comment by alexisju** [comment]
Maybe some style still missing to table (borders to table, tr, td,..)?

In the readme, it can be usefull to give some example how to use markdown, and a link to all possibilities (http://parsedown.org/tests/, or? https://guides.github.com/features/mastering-markdown/)

**Comment by alexisju** [comment]
What about parsing rss feed ? (maybe for later, 0.8.0 ? cfr #273)

**Comment by alexisju** [comment]
The plugin need also to render daily page, now that is possible with https://github.com/shaarli/Shaarli/pull/407/

**Comment by ArthurHoaro** [comment]
Good catch.

**Comment by ArthurHoaro** [comment]
Almost ready to go!

 - [ ] Unit test hooks.
 - [ ] Add a link to markdown doc in `editlink.html` page.
 - [ ] Add a meta file to comply with #378.

RSS feeds will be handled after being refactored.

**Comment by alexisju** [comment]
Currently (without) md plugin, if you write something on two (or more) lines, Shaarli render it as several lines (and put `<br>` between elements). With the plugin, currently, if there is no space or special character at the begining of the line, the outpout is on one line, as one `<p></p>`.

I think it is a bit anoying for the rendering of "old" link description, an not the expected behavior (for example github markdown doen't have this issue).

**Comment by nodiscc** [comment]
With the orginal [markdown](http://daringfireball.net/projects/markdown/) syntax you need 2 spaces at the end of a line, else the linebreak is ignored. Github uses [GFM](https://help.github.com/articles/github-flavored-markdown/).

But I think the plugin should at least not ignore linebreaks, in order to preserve old descriptions formatting,

**Comment by ArthurHoaro** [comment]
You're right. Even though it's an expected behavior - even in [commonmark](http://spec.commonmark.org/0.22/#soft-line-breaks) which I believed the guys at Github were working with, but anyway - it would messed up all previous shaares layout.

I'll run a few test later to fix that.

**Comment by ArthurHoaro** [comment]
Updated.

Additional change: removed hard limit of 400px height for shaares, because style can make them long, and it was annoying anyway.

**Comment by alexisju** [comment]
CSS : need different styles for ol li and ul li

**Comment by ArthurHoaro** [comment]
Thank you, updated. I have to reset lists style because shaarli's CSS mess them up, and I don't want to touch it too much since we plan to rewrite it.

**Comment by ArthurHoaro** [comment]
Moving milestone back to 0.6.3 since the 0.6 is running slow and this PR is probably ready to merge.

**Pull request merged**